### ISSUE TITLE
Resume from sleep with no network results in high CPU

### ISSUE THREAD

**Issue by docsmooth** [abraunegg/onedrive#693]
**Note:** Before submitting a bug report, please ensure you are running the latest 'onedrive' client as built from 'master' and by using the latest available DMD compiler. Refer to the readme on building the client for your system.

### Bug Report Details ###
If OneDrive is synched and running, but then you put your laptop to sleep and wake it with no network connection, the following is logged VERY frequently:
```
Oct 19 22:33:37 prlnscap02 onedrive[22292]: ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?                                                                                                 
Oct 19 22:33:37 prlnscap02 onedrive[22292]: ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object
Oct 19 22:33:37 prlnscap02 onedrive[22292]: ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?                                                                                                 
Oct 19 22:33:37 prlnscap02 onedrive[22292]: ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object
```

This resultant retry and logging logic causes a lot of battery utilization on laptops.  OneDrive client needs some timeout/fallback logic for this error.

**Application and Operating System Details:**
```
$ uname -a && cat /etc/*release*
Linux prlnscap02 5.2.21.20191013 #2 SMP PREEMPT Sat Oct 12 20:00:45 CDT 2019 x86_64 x86_64 x86_64 GNU/Linux
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
DISTRIB_DESCRIPTION="Ubuntu 19.04"
NAME="Ubuntu"
VERSION="19.04 (Disco Dingo)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 19.04"
VERSION_ID="19.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=disco
UBUNTU_CODENAME=disco
```
*   OneDrive Account Type
OneDrive for Business (Office365)
*   DMD or LDC compiler version `dmd --version` or `ldmd2 --version`
```
$ dmd --version
DMD64 D Compiler v2.088.0
Copyright (C) 1999-2019 by The D Language Foundation, All Rights Reserved written by Walter Bright
```
*   Application configuration: Output of `onedrive --display-config`
```
$ onedrive --display-config
onedrive version                    = v2.3.8-4-gaf07dea
Config path                         = /home/rob/.config/onedrive
Config file found in config path    = true
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/rob/Documents/OneDrive-BT
Config option 'skip_dir'            = 
Config option 'skip_file'           = \.*|~*|thumbs.db|*.iso|*.exe|*.sh|*.zip|Thumbs.db*
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 60
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = false
```
*   Curl Version: Output of `curl --version`
```
$ curl --version
curl 7.64.0 (x86_64-pc-linux-gnu) libcurl/7.64.0 OpenSSL/1.1.1b zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh/0.8.6/openssl/zlib nghttp2/1.36.0 librtmp/2.3
Release-Date: 2019-02-06
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
```
**To Reproduce**
1) Have OneDRive syncing normally.
2) Put laptop to sleep and go somewhere else
3) Wake up and check logs

I *believe* that you could just disable your network and get the same, but I haven't tried... yet.

### Bug Report Checklist ###

I *think* that this is caused by this block in onedrive.d:
```
        } catch (CurlException e) {
            // Potentially Timeout was reached on handle error
            log.error("ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?");
            return json;
        }
```
We're catching a known case, but returning json instead of throwing a OneDrive Exception that could be caught higher up in sync.d around line 271: 
```
        } catch (OneDriveException e) {
            if (e.httpStatusCode == 400) {
                // OneDrive responded with 400 error: Bad Request
                log.error("\nERROR: OneDrive returned a 'HTTP 400 Bad Request' - Cannot Initialize Sync Engine");
                // Check this
                if (cfg.getValueString("drive_id").length) {
                    log.error("ERROR: Check your 'drive_id' entry in your configuration file as it may be incorrect\n");
                }
                // Must exit here
                exit(-1);
            }
```
Maybe a sleep and retry, or a full fallback function?

I don't know D yet, but... I'll see if I can patch it myself in the next 2 weeks.

**Comment by abraunegg** [comment]
@docsmooth 
Please update your client version to something that is more current and re-test.

**Comment by docsmooth** [comment]
I did something dumb then, because I last updated 7 days ago after the last commit to master:  a "git pull && make all" results in "alrady up-to-date" and "nothing to be done for 'all'"

00:45 rob@prlnscap02:~/workspace/onedrive$ ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for dmd... dmd
checking version of D compiler... 2.088.0
checking for curl... yes
checking for sqlite... yes
configure: creating ./config.status
config.status: creating Makefile
config.status: creating contrib/pacman/PKGBUILD
config.status: creating contrib/spec/onedrive.spec
config.status: creating onedrive.1
config.status: creating contrib/systemd/onedrive.service
config.status: creating contrib/systemd/onedrive@.service
00:45 rob@prlnscap02:~/workspace/onedrive$ make
make: Nothing to be done for 'all'.
00:45 rob@prlnscap02:~/workspace/onedrive$ ./install-sh
./install-sh: no input file specified.
00:46 rob@prlnscap02:~/workspace/onedrive$ make onedrive
make: 'onedrive' is up to date.
00:46 rob@prlnscap02:~/workspace/onedrive$ make all
make: Nothing to be done for 'all'.
00:46 rob@prlnscap02:~/workspace/onedrive$ git pull
Already up to date.
00:46 rob@prlnscap02:~/workspace/onedrive$ make all
make: Nothing to be done for 'all'.
00:46 rob@prlnscap02:~/workspace/onedrive$ onedrive --version
onedrive v2.3.8-4-gaf07dea
00:46 rob@prlnscap02:~/workspace/onedrive$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
00:46 rob@prlnscap02:~/workspace/onedrive$ git log |head
commit 9d0a6539bd6681af5f078fad939bd63dcdc0b672
Author: Norbert Preining <norbert@preining.info>
Date:   Wed Oct 9 12:28:03 2019 +0900

    Remove unused and unexported SED variable in Makefile.in (#687)

    * Remove unused and unexported SED variable in Makefile.in

commit 6a8d5759412e718abe280004ee00d6359ef9583c
Author: abraunegg <alex.braunegg@gmail.com>

**Comment by abraunegg** [comment]
@docsmooth 
I am able to replicate this even with the latest master:
```
Uploading new items of .
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
ERROR: onedrive.getPathDetailsById call returned an invalid JSON Object
ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object
ERROR: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object
...
```

There is no 'nice way' to have a service loop forever in the background when there is no Internet connectivity.

Working on a patch to update this behavior.

**Comment by abraunegg** [comment]
@docsmooth 
Please can you try the following PR:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/694/head:pr694
git checkout pr694
./configure; make clean; make;
```

With this PR, when Internet connectivity is lost, the following will happen:
```
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
ERROR: OneDrive returned an error with the following message:
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
Exiting application due to Internet connectivity issues
```

If using systemd, systemd will then be leaned on for service restoration based on Internet connectivity.

**Comment by norbusan** [comment]
I somehow disagree with the changes proposed. Why should a sync service **not** run when there is no connectivity available? The fix should not be to kill the server! In particular, since this is the **worst** we can do: It would mean that systemd will restart it - it will fail again - restart - and systemd will after a cetain time mark this service as broken and **disable** it (afair).

What we should fix is the reporting, that is all. And concenring the high CPU usage, I don't see that. I often am in suroundings without network connectivity with my laptop, and onedrive doesn't eat more or less.

So for now I am not overly happy with the current status.

**Comment by abraunegg** [comment]
@docsmooth 
Many thanks for the patch - better than just exiting.

@norbusan 
Thoughts on PR #695 ?

**Comment by abraunegg** [comment]
@docsmooth 
I do not see any high CPU however (Internet stopped via local iptables rule ~6.15am):

![cpuload](https://user-images.githubusercontent.com/4956234/67164972-9974f380-f3cb-11e9-88b0-511b34229c3c.jpg)

The spikes in the graph are when I am compiling / making sure the system is in sync.

**Comment by abraunegg** [comment]
@docsmooth 
Although when unblocking the Internet, the messaging is not correct:
```
Uploading new items of .
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
ERROR: OneDrive returned an error with the following message:
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: Was unable to reconnect to the Microsoft OneDrive service after 10000 attempts lasting over 1.2 years!
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101

```

**Comment by abraunegg** [comment]
@docsmooth , @norbusan 
With the latest changes, this occurs on Internet access resume:
```
The file has not changed
Uploading new items of .
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
ERROR: OneDrive returned an error with the following message:
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
Connectivity to Microsoft OneDrive service has returned
ERROR: onedrive.getPathDetailsById call returned an invalid JSON Object
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

**Comment by abraunegg** [comment]
Updated patch to ensure when retryAttempts == retryCount correct messaging is performed.
```
Uploading new items of .
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
Applying changes of Path ID: 66D53BE8A5056ECA!101
ERROR: OneDrive returned an error with the following message:
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: Was unable to reconnect to the Microsoft OneDrive service after 10000 attempts lasting over 1.2 years!
ERROR: OneDrive returned an error with the following message:
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
Connectivity to Microsoft OneDrive service has returned
ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
```

All of the `ERROR: onedrive.viewChangesByItemId call returned an invalid JSON Object` output will be fixed by #696

**Comment by abraunegg** [comment]
Closing issue as PR #695 merged into master
